duckduckgo/BrowserServicesKit
# Settings Sync Provider with support for syncing Email Protection

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1205254652502303/f
iOS PR: https://github.com/duckduckgo/iOS/pull/1963
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1565
What kind of version bump will this require?: Major

**Optional**:

Tech Design URL: https://app.asana.com/0/481882893211075/1205248329666560/f

**Description**:
Updated SyncMetadata Core Data model with a new version that adds SyncableSettingsMetadata entity.
This entity keeps last modification timestamps for each syncable setting (identified by its Sync key).
Added SettingsProvider that is composed of multiple per-setting providers with their respective handlers.
Handlers conform to SettingsSyncHandling protocol that provides a unified interface for setting and retrieving
values, as well as allows for fine-tuning deduplication logic. Currently there's only one handler – for Email
Protection auth state (EmailProtectionSyncHandler). It own an instance of EmailManager, which means that
now we have multiple (at least 2) EmailManagers that can modify Email Protection state concurrently.
I've added a static lock to EmailManager to avoid data races. Also some of EmailManager APIs got updated
to throw errors instead of only notifying requestDelegate.



**Steps to test this PR**:
* Initial sync: test merge strategy according to [this Asana task](https://app.asana.com/0/0/1205134166546873/f).
* Regular sync: test that updates to Email Protection state are correctly synced to other devices.
* Verify that in-context Email Protection signup triggers sync:
  * disable Email Protection
  * try to sign up to some website and click Dax logo in the email field
  * follow instructions to enable Email Protection
  * verify that Sync is triggered after enabling Email Protection



**OS Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
